omniisaacgymenvs/robots/sensors/proprioceptive/Type.py

```diff
@@ -0,0 +1,202 @@
+        assert len(self.angular_velocity.shape) == 2, f"need to be batched tensor."
+    
+    @staticmethod
+    def quat_to_mat(quat: torch.Tensor) -> torch.Tensor:
+        """
+        Convert batched quaternion to batched rotation matrix.
+        Args:
+            quat (torch.Tensor): batched quaternion.(..., 4)"""
+        EPS = 1e-5
+        w, x, y, z = torch.unbind(quat, -1)
+        two_s = 2.0 / ((quat * quat).sum(-1) + EPS)
+        R = torch.stack(
+        (
+        1 - two_s * (y * y + z * z),
+        two_s * (x * y - z * w),
+        two_s * (x * z + y * w),
+        two_s * (x * y + z * w),
+        1 - two_s * (x * x + z * z),
+        two_s * (y * z - x * w),
+        two_s * (x * z - y * w),
+        two_s * (y * z + x * w),
+        1 - two_s * (x * x + y * y),
+        ),
+        -1,
+        )
+        return R.reshape(quat.shape[:-1] + (3, 3))
+
+    @property
+    def body_transform(self) -> torch.float32:
+        """
+        Return transform from inertial frame to body frame(= inverse of body pose).
+        T[:, :3, :3] = orientation.T
+        T[:, :3, 3] = - orientation.T @ position
+        Returns:
+            transform (torch.float32): transform matrix from inertial frame to body frame.
+        """
+        transform = torch.zeros(self.position.shape[0], 4, 4).to(self.orientation.device)
+        orientation = self.quat_to_mat(self.orientation)
+        transform[:, :3, :3] = orientation.transpose(1, 2)
+        transform[:, :3, 3] = - 1 * torch.bmm(orientation.transpose(1, 2), self.position[:, :, None]).squeeze()
+        return transform
+
+@dataclasses.dataclass
+class ImuState:
+    """
+    IMU state typing class. 
+    Args:
+        angular_velocity (torch.float32): angular velocity of the body in body frame.
+        linear_acceleration (torch.float32): linear acceleration of the body in body frame.
+    """
+
+    angular_velocity: torch.float32 = torch.zeros(1, 3)
+    linear_acceleration: torch.float32 = torch.zeros(1, 3)
+
+    def update(self, angular_velocity:torch.float32, linear_acceleration:torch.float32) -> None:
+        """
+        Update internal attribute from arguments.
+        Args:
+            angular_velocity (torch.float32): angular velocity of the body in body frame.
+            linear_acceleration (torch.float32): linear acceleration of the body in body frame.
+        """
+        self.angular_velocity = angular_velocity
+        self.linear_acceleration = linear_acceleration
+    
+    def reset(self, num_envs:int) -> None:
+        """
+        Reset internal attribute to zero.
+        """
+        self.angular_velocity = torch.zeros(num_envs, 3)
+        self.linear_acceleration = torch.zeros(num_envs, 3)
+    
+    def reset_idx(self, env_ids:torch.Tensor) -> None:
+        """
+        Reset internal attribute of specified env to zero.
+        """
+        self.angular_velocity[env_ids] = 0
+        self.linear_acceleration[env_ids] = 0
```

Comment:
Is reset supposed to reset all environments while reset_idx just some of them? It might be better to have only reset_idx, and pass all the indices to reset_idx.